src/funnels/SwapDepositor.sol

```diff
@@ -0,0 +1,222 @@
+         )))));
+    }
+
+    struct SwapDepositParams {
+        address gem0;
+        address gem1;
+        uint256 amt0;
+        uint256 amt1;
+        uint256 minAmt0;
+        uint256 minAmt1;
+        uint256 minSwappedOut;
+        uint24 fee;
+        int24 tickLower;
+        int24 tickUpper;
+        address swapperCallee;
+        bytes swapperData;
+    }
+
+    function sortDepositedTokens(SwapDepositParams memory p) internal pure returns (SwapDepositParams memory) {
+        (p.gem0, p.gem1, p.amt0, p.amt1, p.minAmt0, p.minAmt1) 
+            = p.gem0 < p.gem1 ? (p.gem0, p.gem1, p.amt0, p.amt1, p.minAmt0, p.minAmt1) : (p.gem1, p.gem0, p.amt1, p.amt0, p.minAmt1, p.minAmt0);
+        return p;
+    }
+
+    function getConvertedAmounts(SwapDepositParams memory p) internal view returns (uint256 amt0_1, uint256 amt1_0) {
+        (uint160 sqrtPriceX96,,,,,,) = UniV3PoolLike(getPoolAddress(p.gem0, p.gem1, p.fee)).slot0();
+        uint256 p0 = uint256(sqrtPriceX96) * uint256(sqrtPriceX96) * WAD / uint256(2 ** 192);
+        amt0_1 = p.amt0 * p0 / WAD;
+        amt1_0 = p.amt1 * WAD / p0;
+    }
+
+    function swapBeforeDeposit(SwapDepositParams memory p) internal returns (uint256 amt0, uint256 amt1) {
+        (uint256 amt0_1, uint256 amt1_0) = getConvertedAmounts(p);
+        // TODO2: calculate more accurate swap amounts when callee is UniV3SwapperCallee
+        
+        uint bought;
+        uint sold;
+        if (p.amt0 > amt1_0 + gaps[p.gem0]) {
+            // need to sell some gem0
+            sold = (p.amt0 - amt1_0) / 2;
+            bought = Swapper(swapper).swap(p.gem0, p.gem1, sold, p.minSwappedOut, p.swapperCallee, p.swapperData);
+            amt0 = p.amt0 - sold;
+            amt1 = p.amt1 + bought;
+
+        } else if (p.amt1 > amt0_1 + gaps[p.gem1]) {
+            // need to sell some gem1
+            sold = (p.amt1 - amt0_1) / 2;
+            bought = Swapper(swapper).swap(p.gem1, p.gem0, sold, p.minSwappedOut, p.swapperCallee, p.swapperData);
+            amt1 = p.amt1 - sold;
+            amt0 = p.amt0 + bought;
+        } else {
+            amt1 = p.amt1;
+            amt0 = p.amt0;
+        }
+    }
+
+
+    function deposit(SwapDepositParams memory p) external auth returns (uint128 liquidity, uint256 amt0, uint256 amt1) {
+        p = sortDepositedTokens(p);
+        (amt0, amt1) = swapBeforeDeposit(p);
+        Depositor.DepositParams memory depositParams = Depositor.DepositParams({
+            gem0: p.gem0,
+            gem1: p.gem1,
+            amt0: amt0,
+            amt1: amt1,
+            minAmt0: p.minAmt0,
+            minAmt1: p.minAmt1,
+            fee: p.fee,
+            tickLower: p.tickLower,
+            tickUpper: p.tickUpper
+        });
+        (liquidity, amt0, amt1) = Depositor(depositor).deposit(depositParams);
+    }
+
+    struct WithdrawSwapParams {
+        address gem0;
+        address gem1;
+        uint128 liquidity;
```

Comment:
I know this is in the nft manager (and the pool) interface, but not sure we can expect the keeper/facilitator to supply an amount of "liquidity" when needing to withdraw. He probably doesn't know how to calculate that easily.

Reply:
Resolving as we'll do an automated depositor only.